### PR TITLE
Add rate limiter and update tests.

### PR DESCRIPTION
**Pull request by howbazaar**
Add configurable rate limiting for agent connections.

This work took a sideways step as I needed to write tests for the rate limiting, but all the login tests were using JujuConnSuite. So I changed the login tests to instead be based on statetetesting.StateSuite, which is much lighter.

## QA steps

```sh
juju bootstrap lxd test
juju controller-config agent-ratelimit-max=5
juju controller-config agent-ratelimit-rate=500ms
juju deploy ~jameinel/ubuntu-lite -n 5
for i in {1..5}; do juju add-unit ubuntu-lite -n 5 --to 0,1,2,3,4; done
watch -c -n 1 juju status --color
# in a different shell 
juju ssh -m controller 0
sudo service jujud-machine-0 restart
# observe the reduced rate of reconnection in the status 
```

## Documentation changes

Will need to document the two new controller config values.

### CODE REVIEW

**Comment by howbazaar** [comment]
I need to update this to support disabling the ratelimiting.

**Comment by hpidcock** [comment]
LGTM. Works well, tested the rate limiting on microk8s with ~50 agents.

File: state/controller_test.go [hunk 1]
```diff
@@ -29,29 +29,31 @@ func (s *ControllerSuite) TestControllerAndModelConfigInitialisation(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 
 	optional := set.NewStrings(
-		controller.IdentityURL,
-		controller.IdentityPublicKey,
+		controller.AgentRateLimitMax,
```
Comment (jameinel) [inline state/controller_test.go]: I assume this is just Sort() the list alphabetically

File: apiserver/apiserver.go [hunk 2]
```diff
@@ -299,6 +302,25 @@ func newServer(cfg ServerConfig) (_ *Server, err error) {
...
+		controllermsg.ConfigChanged,
+		func(topic string, data controllermsg.ConfigChangedMessage, err error) {
+			if err != nil {
+				logger.Criticalf("programming error in %s message data: %v", topic, err)
+				return
+			}
+			srv.updateAgentRateLimiter(data.Config)
+		})
+	if err != nil {
+		logger.Criticalf("programming error in subscribe function: %v", err)
```
Comment (jameinel) [inline apiserver/apiserver.go]: Criticalf in this code.

Comment (howbazaar) [reply apiserver/apiserver.go]: This one is critical by design.

File: juju/testing/conn.go [hunk 3]
```diff
@@ -534,6 +534,8 @@ func (s *JujuConnSuite) setUpConn(c *gc.C) {
 	for key, value := range s.ControllerConfigAttrs {
 		s.ControllerConfig[key] = value
 	}
+	// Explicitly disable rate limiting.
+	s.ControllerConfig[controller.AgentRateLimitMax] = 0
```
Comment (hpidcock) [inline juju/testing/conn.go]: Nice

File: state/controller_test.go (hunk 1 above)
Comment (howbazaar) [reply state/controller_test.go]: Yeah, I hit my OCD limit

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$merge$$

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**